packages/core/src/sync-realtime/index.ts

```diff
@@ -571,44 +687,70 @@ export const createRealtimeSync = (
 
     // Remove logs that don't match a filter, recording required transactions
     logs = logs.filter((log) => {
-      let isLogMatched = false;
+      let isMatched = false;
 
       for (const filter of logFilters) {
         if (isLogFilterMatched({ filter, block, log })) {
-          isLogMatched = true;
+          // TODO: includeTransactionReceipt
+          // return early if includeTransactionReceipt, otherwise continue checking filters
           requiredTransactions.add(log.transactionHash);
-          if (filter.includeTransactionReceipts) {
-            requiredTransactionReceipts.add(log.transactionHash);
-          }
+          isMatched = true;
         }
       }
 
-      return isLogMatched;
+      return isMatched;
     });
 
-    // Remove call traces that don't match a filter, recording required transactions
-    callTraces = callTraces.filter((callTrace) => {
-      let isCallTraceMatched = false;
-      for (const filter of callTraceFilters) {
-        if (isCallTraceFilterMatched({ filter, block, callTrace })) {
-          isCallTraceMatched = true;
-          requiredTransactions.add(callTrace.transactionHash);
-          if (filter.includeTransactionReceipts) {
-            requiredTransactionReceipts.add(callTrace.transactionHash);
-          }
+    // Initial weak trace filtering before full filtering with factory addresses in handleBlock
+    traces = traces.filter((trace) => {
+      let isMatched = false;
+      for (const filter of transferFilters) {
+        if (
+          isTransferFilterMatched({
+            filter,
+            block: { number: block.number },
+            trace: trace.trace,
+          })
+        ) {
+          // TODO: includeTransactionReceipt
+          // return early if includeTransactionReceipt, otherwise continue checking filters
+          requiredTransactions.add(trace.transactionHash);
+          isMatched = true;
         }
       }
 
-      return isCallTraceMatched;
+      for (const filter of traceFilters) {
+        if (
+          isTraceFilterMatched({
+            filter,
+            block: { number: block.number },
+            trace: trace.trace,
+          })
+        ) {
+          // TODO: includeTransactionReceipt
+          // return early if includeTransactionReceipt, otherwise continue checking filters
+          requiredTransactions.add(trace.transactionHash);
+          isMatched = true;
+        }
+      }
+
+      return isMatched;
     });
 
     ////////
     // Transactions
     ////////
 
-    const transactions = block.transactions.filter(({ hash }) =>
-      requiredTransactions.has(hash),
-    );
+    const transactions = block.transactions.filter((transaction) => {
```

Comment:
My understanding is that the `fetchBlockEventData` function should be very conservative with filtering out data, because only the `handleBlock` function has access to the newest child addresses.

I'm not quite able to follow whether or not we're accomplishing that as expected here, so worth a double check. Seems like a plausible way to introduce a missing event regression that our tests might not catch.

@kyscott18 and @khaidarkairbek  I think once we're ready to ship `0.8`, we should do one final comprehensive review of these two functions.

Reply:
Yes, it is pretty complex. I've reread it a couple of times. We want to filter out as much as possible so that dependent requests (`eth_getTransactionReceipt`) can be skipped. Otherwise we could filter out no data in `fetchBlockEventData` and only in `handleBlock`